python/hidet/graph/ops/reduce/reduce.py

```diff
@@ -48,69 +53,110 @@ def allow_epilogue(self) -> bool:
             # use self.cuda_schedule_reduce_by_default
             return False
 
+    def allow_prologue(self) -> bool:
+        return False
+
     def implement_cuda(self, working_dir: str) -> IRModule:
         rank = len(self.inputs[0].shape)
         if rank - 1 in self.dims:
-            return self.cuda_schedule_reduce_by_warp()
+            return tune.extract_ir_modules(self.cuda_schedule_reduce_by_warp)
         else:
             return self.cuda_schedule_reduce_by_default()
 
-    def cuda_schedule_reduce_by_warp(self) -> IRModule:
+    @tune.space(2, use_atomic=[True, False])
+    @tune.space(1, use_atomic=[True, False])
+    def cuda_schedule_reduce_by_warp(self, use_atomic=True) -> IRModule:
         import hidet
         from hidet.ir.primitives import active_mask, shfl_down_sync, shfl_sync
         from hidet.ir.compute import ReduceOperation
         from hidet.ir.type import data_type, Int
         from hidet.ir.layout import row_major
-        from hidet.lang import spatial, repeat, attrs, cast
-        from hidet.lang.cuda import blockIdx, threadIdx
+        from hidet.lang import spatial, repeat, attrs, cast, tensor_pointer, address
+        from hidet.lang.cuda import dynamic_shared_memory, syncthreads
 
-        warp_size = 32
-        block_size = warp_size
         x, y = self.inputs[0], self.outputs[0]
+        warp_size = 32
         xdtype = x.type.dtype
         shape: List[Int] = list(x.shape)
+        lanes = 1
+        vtype: DataType = xdtype
+        if xdtype.nbytes < 4:
+            num_eles: int = 4 // xdtype.nbytes
+            if shape[-1] % num_eles == 0:
```

Comment:
consider adding an `assert is_constant(shape[-1])` if this schedule template only supports the case where the last dimention is known.

Reply:
Or change it to
```python
if is_constant(shape[-1]) and shape[-1] % num_eles == 0:
    ...
```